### PR TITLE
MattT/CASEFLOW-4600 Add Assigned Tab to EMO Queue

### PR DESCRIPTION
**Pull request by ThorntonMatthew**
Resolves [CASEFLOW-4600](https://vajira.max.gov/browse/CASEFLOW-4600)

### Description
Adding an 'Assigned' tab to Caseflow Queue for the EMO organization.

### Acceptance Criteria
- [ ] Code compiles correctly
- [ ] Determination of the behavior and triggers of the Assigned tab
- [ ] When cases are assigned to one of the three (3) RPO organizations, they appear on the Assigned tab
- [ ] If the RPO organization sends it to board intake, the appeal remains in EMO’s Assigned tab (it is also moved to the Completed tab of the RPO queue)
- [ ] If the RPO organization sends the appeal back to the EMO office, it is removed from the Assigned tab goes to the Unassigned tab
- [ ] If the EMO organization sends an appeal to BVA Intake, regardless of the reason selected (not an education issue or appeal ready for review), the appeal shows on the EMO's "Assigned" tab.

### Testing Plan
1. Ensure that you have the `edu_predocket_appeals` feature toggle enabled. You can paste `FeatureToggle.enable!(:edu_predocket_appeals)` into the Rails console in order to enable it.
2. If you have not yet reseeded your database with the seed file updates from the feature branch, run `make seed-dbs`. The updated seed files are present in this branch.
3. Open Caseflow and switch to an Intake user.
4. Navigate to Caseflow Intake and begin the 10182 entry flow.
5. Whenever you arrive at the Add/Remove issues page, add an issue with the benefit type of 'Education' and select 'Yes' in the pre-docketing radio field.
6. Submit the appeal.
7. Switch to an EMO user.
8. Navigate to the EMO Organization's Queue. It should be [at this address](http://localhost:3000/organizations/edu-emo?tab=education_assigned&page=1).
9. Ensure that the appeal is **not present**. Whenever an appeal is first assigned to the EMO it will be present in the 'Unassigned' tab until an action is taken upon it.
10. Open up your Rails console.
11. Run `a = Appeal.order("created_at").last` to get the appeal you've just submitted.
12. Due to the lack of full UI functionality at this time, you must manually edit the status of the appeal's task(s) in order to get it to appear in the Assigned tab. I encourage you to ensure that the appeal appears in the tab with all three statuses:

```
a.tasks.last.update!(status: Constants.TASK_STATUSES.on_hold)
a.tasks.last.update!(status: Constants.TASK_STATUSES.completed)
a.tasks.last.update!(status: Constants.TASK_STATUSES.cancelled)
```
13. To simulate a scenario where the EMO has sent an appeal to an RPO, enter the following commands. Ensure that the task remains in the EMO's assigned tab afterwards.
```
a.tasks.last.update!(status: Constants.TASK_STATUSES.on_hold)

EducationAssessDocumentationTask.create!(parent: a.tasks.last, appeal: a, assigned_at: Time.zone.now,assigned_to: EduRegionalProcessingOffice.first)
```

14. Set the `EducationDocumentSearchTask` to `completed`.

`a.tasks.last.update!(status: Constants.TASK_STATUSES.completed)`

15. If you simulated adding an RPO task, you can set that to be completed as well:

`a.tasks.last.children.first.update!(status: Constants.TASK_STATUSES.completed)`

16. Set the parent, `PreDocketTask` (BVA Intake), to a status of completed. Ensure that the appeal is no longer present in the Assigned tab afterwards (it will move to the completed tab). This simulates BVA Intake sending a pre-docket appeal to docket.

`a.tasks.last.parent.first.update!(status: Constants.TASKS_STATUSES.completed)`

17. Now set the  `PreDocketTask` to a status of `cancelled`. This simulates BVA Intake removing all of the issues from an appeal to cancel it. The appeal should not appear in the Assigned tab once you have done so.
18. To simulate an RPO sending an appeal back to EMO, you can run the following commands. The appeal should no longer appear in the Assigned tab (it would be in the in the unassigned tab).
```
# If you have adjusted the status of the Pre Docket Task, set it back to on hold
a.tasks.last.parent.first.update!(status: Constants.TASKS_STATUSES.on_hold)

# Set the RPO task to completed if it isn't already.
a.tasks.last.children.first.update!(status: Constants.TASK_STATUSES.completed)

# Set EMO task to status: assigned, which is the status it will have if an RPO sends an appeal to EMO.
a.tasks.last.update!(status: Constants.TASK_STATUSES.assigned)
```

- [ ] For higher-risk changes: [Deploy the custom branch to UAT to test](https://github.com/department-of-veterans-affairs/appeals-deployment/wiki/Applications---Deploy-Custom-Branch-to-UAT)

### User Facing Changes
 - [ ] Screenshots of UI changes added to PR & Original Issue

 There is no before image available for this change as it is a brand new UI addition.

Edu EMO Assigned Tab

### Code Documentation Updates
- [ ] Add or update code comments at the top of the class, module, and/or component.

### CODE REVIEW

File: app/models/queue_tabs/education_emo_assigned_tasks_tab.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
+# frozen_string_literal: true
+
+class EducationEmoAssignedTasksTab < QueueTab
+  validate :assignee_is_organization
+
+  attr_accessor :show_reader_link_column, :allow_bulk_assign
```
Comment (ajspotts) [inline app/models/queue_tabs/education_emo_assigned_tasks_tab.rb:6]: @ThorntonMatthew do we need either of these `attr_accessor`s? I believe we're only showing the reader column in the Unassigned tab according to the designs and I don't think we're handling bulk assign in any context of education pre-docket (I know we were for VHA)

File: app/models/queue_tabs/education_emo_assigned_tasks_tab.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+  def description
+    format(COPY::EDUCATION_EMO_QUEUE_PAGE_ASSIGNED_TASKS_DESCRIPTION, assignee.name)
+  end
+
+  def task_ids_where_parent_has_been_closed
+    closed_tasks.select { |task| task.parent.closed? }.pluck(:id)
+  end
+
+  def task_ids_emo_sent_elsewhere
```
Comment (ajspotts) [inline app/models/queue_tabs/education_emo_assigned_tasks_tab.rb:24]: Could we potentially make this method slightly more descriptive as far as what's actually going on? If these are appeals assigned to RPOs & BVA Intake perhaps something like `tasks_assigned_to_rpos_or_bva` or something similar

File: spec/models/queue_tabs/education_emo_assigned_tasks_tab_spec.rb [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            :in_progress,
+            parent: task,
+            assigned_to: regional_processing_office
+          )
+          task.update!(status: Constants.TASK_STATUSES.on_hold)
+          task.children
+        end.flatten
+      end
+
+      it "the apepal appears in the EMO's assigned tab while the RPO works on the appeal" do
```
Comment (ajspotts) [inline spec/models/queue_tabs/education_emo_assigned_tasks_tab_spec.rb:67]:
```suggestion
      it "the appeal appears in the EMO's assigned tab while the RPO works on the appeal" do
```

File: app/models/queue_tabs/education_emo_assigned_tasks_tab.rb (hunk 1 above)
Comment (ThorntonMatthew) [reply app/models/queue_tabs/education_emo_assigned_tasks_tab.rb]: These are not needed for this tab. I have removed them.

File: app/models/queue_tabs/education_emo_assigned_tasks_tab.rb [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    closed_tasks.select { |task| task.parent.closed? }.pluck(:id)
+  end
+
+  def task_ids_assigned_to_rpos_or_bva
+    on_hold_tasks.map(&:id) + closed_tasks.map(&:id)
+  end
+
+  def tasks
+    Task.includes(*task_includes).visible_in_queue_table_view.where(
+      id: (task_ids_emo_sent_elsewhere - task_ids_where_parent_has_been_closed)
```
Comment (ajspotts) [inline app/models/queue_tabs/education_emo_assigned_tasks_tab.rb:28]: @ThorntonMatthew think you'll need to update this line as well to reflect the new name

**Review by ajspotts — APPROVED** [review]
lgtm!

**Pull request merged**